obspy/signal/src/evalresp/evalresp.1
.\" This program has been completely rewritten from the original version authored by Jean Francois Fels
.\" to support several new features.  Among the new features supported are
.\" (a) a "new" RESP file format that contains the blockette and$
.\" field numbers as prefixes to each line.  This allows for$
.\" quick determination of whether or not the program is$
.\" parsing the correct information without relying on searching$
.\" for non-standardized character strings in the RESP file$
.\" (b) support for the blockette [61] responses$
.\" (c) support for the response-reference style responses (i.e.$
.\" a blockette [60] followed by a series of blockette [41] or$
.\" blockette [43] through blockette [48] responses)$
.\" Author: Thomas J. McSweeney
.\" Phone: (206) 547-0393
.\" Current support:	ISTI
.\" Internet: info@isti.com
.\" Phone: (518) 602-0001
.\" Also: rick@iris.washington.edu
.\" Phone: (206) 547-0393
.\" 
.TH "EVALRESP" "V3.3.3" "03-May-2010" "" "IRIS programs"
.SH "NAME"
\fIevalresp\fR \- evaluate response information and output to ASCII files using \fIrdseed V4.16\fR and above RESP files.
.SH "SYNOPSIS"
evalresp STA_LIST CHA_LIST YYYY DAY MIN_FREQ MAX_FREQ NFREQS [\fB\-f\fR file] [\fB\-u\fR units]
[\fB\-t\fR time\-of\-day] [\fB\-s\fR type\-of\-spacing] [\fB\-r\fR resp\-type] [\fB\-n\fR network\-id]
[\fB\-l\fR location\-id] [\fB\-stage\fR start [stop]] [\-stdio] [\-use\-estimated\-delay] [\-unwrap] [\-ts]
[\-il] [\-ii] [\fB\-it\fR tension] [\-v]
.PD 0.3

.SH "DESCRIPTION"
.LP 
\fIEvalresp \fR will calculate the complex response of a specified station or set
of stations, channel or channels, and network, for a specified date, time and frequency, all by using
the SEED ASCII response files produced by the program \fI`rdseed'\fR.
.I Evalresp
by itself returns the usage or `help' lines useful for how to correctly input
the options.
.sp
.SH "COMMAND-LINE PARAMETERS"
.nf 
 \-f file              directory\-name|filename
 \-u units             'dis'|'vel'|'acc'|'def'
 \-t time\-of\-day       HH:MM:SS
 \-s type\-of\-spacing   log|lin
 \-n netid             'II'|'IU'|'G'|'*'...
 \-l locid             '01'|'AA,AB,AC'|'A?'|'*'...
 \-r resp_type         'ap'=amp/pha|'cs'=complex spectra|'fap'=freq/amp/pha
 \-stage start [stop]  start and stop are integer stage numbers
 \-stdio               take input from stdin, output to stdout
 \-use\-estimated\-delay use estimated delay in computation of response
 \-il                  interpolate List blockette output
 \-ii                  interpolate List blockette input
 \-it tension          tension for List blockette interpolation
 \-unwrap              unwrap phase if the output is AP (amplitude\-phase)
 \-ts                  use total sensitivity from stage 0 instead of computed
 \-v                   verbose; list parameters on stdout
.fi 
.sp
.SH "WHAT IS NEW"
\fBSee ChangeLog file from evalresp distribution\fR
.sp
.SH "SUPPORT OF PRESSURE and MAGNETIC DATA"
\fIevalresp\fR is able to process responses of
pressure instruments (in Pascals) and magnetic flux density (in Tesla). Note that the default units 
are assumed to be Pascals or Tesla respectively, therefore the "\-u" option is ignored while
processing pressure and magnetic flux density responses.
.sp
.SH "SEED FILTER COEFFICIENTS AND STAGE DELAY CORRECTION"
As of July 2007 the SEED standard was clarified to require that digital filter coefficients are to be listed in forward order.  For a mathematical description see the SEED manual. As a reference, minimum\-phase filters (which are asymmetric) should be written with the largest values near the beginning of the coefficient list.  Additionally the sign convention for the estimated delay and correction applied specified in Blockette 57 for a given stage were clarified.  In almost all cases both the estimated delay and correction applied should be positive values and very nearly the same, if not exactly, the same value.

Proper evalresp operation requires the conventions described above to be followed.  Unfortunately there are some SEED response descriptions which do not follow the convention.  Using such responses with evalresp will likely result in improper response calculation.  Furthermore, applying improper evalresp results to time series data can result in either obvious, or more problematic, subtle data modifications.
.sp
.SH "FIR FILTERS"
All FIR filters are considered as having a zero phase\-shift, even
if they are not symmetrical and the delay correction is null.
If there are 2 FIR filters in the same stage, the
program assumes that both filters have the same input sample
interval (in other words, the first filter has a decimation factor
of 1). Typically if two FIR filters appear in the same stage,
the second FIR filter is a continuation of the first. This often
results when the FIR filter in question has more than 415 coefficients
(which causes the length of the blockette containing the response
information to exceed the "%4.4d" format of the blockette length specifier
(defined by SEED). When this occurs, a second (continuation) blockette
is written that contains the coefficients that would not fit in the first blockette.
\fIevalresp\fR will handle such continuation filters by joining all FIR filters in the same
stage into one large FIR filter in the order that they were scanned.
.sp
.SH "IIR FILTERS"
Versions 3.2.17 and above support IIR digital filters in coefficients format with a non\-zero phase shift.
IIR coefficients for a single stage must fit in a single blockette.
.sp
.SH "GENERIC RESPONSE BLOCKETTES"
Versions 3.2.17 and above support generic response blockette (SEED blockettes 55).
Generic response blockette is a list of phases and amplitudes computed for the
preselected set of frequencies. This filter type is supported only if the response input file
contains blockette(s) 55 as a stage 1 and possibly channel sensitivity blockette as a stage 0.
If a generic response blockette is recognized in the input, \fIevalresp\fR ignores the user\-defined
frequency sampling from the command line. The ouput, therefore, contains
responses for only those frequencies which have been defined in the generic response blockette.
.sp
.SH "FILTER SEQUENCE"
The program assumes that the response information consists of a series
of filter stages arranged in a cascade.  It is assumed that the first filter
in a given stage is one of the following:  (1) A Laplace\-Transform or Analog pole\-
zero filter, (2) an IIR pole\-zero filter, (3) a FIR filter (either symmetric
or asymmetric), or (4) a stand\-alone gain blockette that indicates the overall
sensitivity of the filter sequence (a stage zero filter). Versions
of \fIevalresp 3.2.17\fR and higher also support (5) IIR digital coefficients filters and (6) provide
limited support for Generic Response Blockette.    It is further assumed that the filters will be followed by
a gain blockette (except Generic Response Blockettes).  If the stage is a decimation stage, then a decimation
blockette will be included. This decimation blockette typically precedes the gain blockette for the stage in a
SEED response file, although the order of the blockettes within a stage does not matter.
If the blockettes within a stage are not in the order that
evalresp expects to find them in, evalresp will rearrange them so that they appear in the "correct" order.
If the response is a single stage response, \fIevalresp\fR will allow the user to specify an overall (stage 0)
gain, rather than requiring the user to specify a stage 1 and stage 0 gain blockette (since, in this case,
the stage 0 and stage 1 gains are identical).
.sp
The stage sequence number is checked by \fIevalresp\fR during parsing and any break in the sequence is
considered to be an error. The result is that filter sequences with out of order stages are rejected as
invalid responses. In addition, the output units of a stage and the input units of the next stage are
compared by \fIevalresp\fR. If the output units of a stage do not match the input units of the next stage, the
filter sequence is considered to be invalid and the response is rejected as an invalid response. The only
exception to this rule are so called "gain\-only" stages. Since these stages have no units associated with
them, the \fIevalresp\fR program will skip them in determining the input units of the next stage. If a gain\-only
filter is found in the sequence, \fIevalresp\fR will scan to the next non\-gain\-only stage and compare the
output units of the current stage with the input units of that stage. Again, a difference in the units will be
considered to be an error in the filter sequence and cause that response to be rejected as invalid.
.sp
.SH "UNEXPECTED CASES:"
\- stand alone FIR filters (i.e. those with no sample rate and gain specified) are discarded.
(Only that stage is discarded, the rest of the filter sequence is kept and used to calculate
a response).
  \- FIR filters which are not normalized to 1 at frequency 0 are normalized.
  \- IIR coefficients filter with a stage containing more than a single blockette 54.
  \- Mixing generic response stage with the other responses in a single file.

.fi 
.SH "HOW THE PROGRAM SEARCHES FOR RESPONSES"
If the `\fB\-f\fR' option is specified, a determination is made as to whether the filename that follows
the `\fB\-f\fR' flag is a directory.
.HP 4
(1) If it is a directory, then that directory, and only that directory, is searched for files with names
like RESP.NET.STA.LOC.CHA (or RESP.NET.STA.CHA), where the NET, STA, and CHA match the user supplied
(or default) network\-code, station names (from the STA_LIST), location\-code, and channel names (from
the CHA_LIST).
.HP 4
(2) If it is not a directory, then a file with that name is used as input to the program. That file, and
only that file, will be searched for response information that matches the user's request.
.HP 4
(3) If the \fB\-f\fR option is not specified, then both the current working directory and the directory pointed
to by the SEEDRESP environment variable (if it exists) are searched for response information
that matches the user's request. As in the directory search (above), the filenames are
constructed automatically. The files are searched starting with the local directory, so if a match
is found in both the local and SEEDRESP directories, the information from the local file will be
used.
.HP 4
(4)  Because it is possible to use wildcards to specify the network\-code, stations and channels that
are of interest, when the \fB\-f\fR flag is used to pass the name of a directory to search or when the \fB\-f\fR
option is not given and the local and SEEDRESP directories are searched for matching files, all
files whose names match the user's requested station, channel, and network code are searched
for responses that have an effective time that includes the requested date (and time, if
specified). This is necessary because there may be multiple, unique station\-channel\-network's
that match a single input station\-channel\-network tuple from the user if wildcards are used. A
list of all of the files that match is constructed and each is searched in turn. However, only the
first matching response in each file is calculated.
.sp
.HP 4
If the \fB\-stdio\fR option is given, the SEED response information is scanned from standard input and
the resulting response is returned to standard output. In this case, the program will continue to
search standard input for matching responses as long as it remains open (i.e. until an EOF is
signaled). This allows the user to place evalresp into a pipeline of commands, or to use I/O
redirection to read SEED responses from a file containing the response information.



.SH "NOTES ABOUT USAGE"
.HP 4
(1)  First, you must create an ASCII file containing the response information for the SEED volume.
For \fIevalresp V3.0\fR (and later), \fIrdseed V4.16\fR or later must be used to create these files. To create
the files, the R option to rdseed can be specified (either on the command line or interactively).
This places the response information in the SEED volume into ASCII files with names like
RESP.NET.STA.LOC.CHA. Alternatively, the \fB\-d\fR option can be specified and, by responding "yes" to
the query of whether you want response files written, these same files will be extracted only for
the station\-channel\-network tuples for which data is extracted from the SEED volume.
.HP 4
(2) If the file argument is a directory, that directory will be searched for RESP files of the form
RESP.NET.STA.LOC.CHA (or RESP.NET.STA.CHA).
.HP 4
(3) If the file argument is a file, that file is assumed to be a concatenated version of the output from
a call to rdseed with the \fB\-R\fR option. If this is the case, then only this file will be searched for
matching response information
.HP 4
(4) If the file argument is missing, the current directory will be searched for RESP files of the form
RESP.NET.STA.LOC.CHA or RESP.NET.STA.CHA (see \fI"How the Program Searches for Responses"\fR, above).
.HP 4
(5) If the environment variable SEEDRESP exists and is the name of a directory, that directory will
also be searched for the requested files (if the \fB\-f\fR option is not used, see \fI"How the Program
Searches for Responses"\fR, above).
.IP 
.sp
i.e. if typed setenv SEEDRESP /foo/resp_dir and no file or directory is specified
to search on the command line, then the current directory and the directory
/foo/resp_dir will be searched for matching RESP files from which to calculate
responses.
.sp
.HP 4
(6) The units argument is one of the following: DIS (displacement), VEL (velocity), ACC
(acceleration), DEF (default units), and represents the units for which the output response
should be calculated (regardless of the units that are used to represent the response in the
RESP file). If Default Units are chosen, the response is calculated in output units/input units,
where these units are exactly the input units of the first stage of the response and the output
units of the last stage of the response. This is a useful alternative if the units for a particular
type of sensor (e.g. a pressure sensor) are not in units that can be converted to displacement,
velocity, or acceleration. The default value for this argument is VEL.
.HP 4
(7) The time\-of\-day argument is in HH:MM:SS format. This is used only in the case where there is
more than one response in a given SEED volume for a given day. In that case, this argument can
be used to choose one response over another according to the effective time of each. If this
argument is not specified, then the first response that is found in the file that matches the
requested year and day will be used. The default value for this argument is 00:00:00.0.
.HP 4
(8) The type\-of\-spacing argument is either logarithmic or linear ("log" or "lin" respectively). This
governs whether the frequencies chosen are spaced evenly between the minimum frequency and
the maximum frequency in a linear or logarithmic sense. This argument defaults to a value of
"log".
.HP 4
(9) The \fB\-v\fR argument indicates that the user would like to receive the verbose ouput from the
\fIevalresp\fR program. When this flag is included on the command line, diagnostic information will be
sent to standard output showing summary information of the calculated response for each
station\-channel\-network tuple that matches the user's request. If this option is not specified,
only error output will occur in the program.
.HP 4
(10) The \fB\-r\fR argument indicates the response type the user desires. Available values are "cs" for
complex\-spectra output, "ap" for amplitude\-phase output, and "fap" for frequency\-amplitude\-phase output.

CHANGELOG.md
master
======

*Changes*

- obspy.core:
  * add option to suppress evalresp sensitivity mismatch warning when removing
    instrument response (see #2677)
  * round magnitudes in Catalog/Event string representation to one decimal
    (see #2742)
  * add support for pathlib.Path objects in read(), read_inventory() and
    read_events() functions (see #2743)
- obspy.clients.fdsn:
  * introduce fine-grained FDSN client exceptions (see #2653)
- obspy.clients.filesystem:
  * add get_waveforms_bulk() method to SDS client (see #2616, #2626)
- obspy.io.css:
  * open CSS waveforms even if gzip-compressed (see #2736)
- obspy.io.hypodd
  * add PHA write support (see #2687)
  * add read support for horizontal and vertical origin uncertainty (see #2687)
- obspy.io.reftek:
  * enable reading data with floating point sampling rates like low sampling
    rate state-of-health channels (see #2678)
  * fix reading data in '16' and '32' encodings, when packets do not use
    the fixed maximum amount of available number of samples per data packet
    (see #2678)
  * properly take into account native system byteorder, should fix reading
    rt130 data on big endian systems (see #2678)
- obspy.io.xseed:
  * fix a bug reading SEED blockettes 48 and 58 which was likely never
    encountered (see #2668)
- obspy.signal.array_analysis
  * fixed an issue in array_processing function returning wrong times
    for matplotlib versions >= 3.3 due to the epoch change in matplotlib
    (see #2723)
- obspy.signal.cross_correlation:
  * Remove deprecated xcorr function, remove deprecated domain keyword
    argument in correlate function (see #1979)
- obspy.signal.spectral_estimation.PPSD:
  * Added special handling option for infrasound data and global infrasound noise
    models for plotting (see #2740)

maintenance_1.2.x
=================

*Changes*

- obspy.core:
  * Inventory addition now consistently uses shallow copies (#2675, #2694)
- obspy.clients.fdsn:
  * add URL mapping for IRISPH5 (see #2739)
- obspy.clients.seedlink:
  * Fix a bug in basic client when printing debug output (see #2734)
- obspy.io.gse2:
  * When reading GSE2 bulletins, station magnitudes now include waveform IDs
    and have associated station magnitude contributions (see #2718)
- obspy.io.xseed:
  * Properly read a given value of 0.0 in station elevation and not replace it
    with bogus value (see #2763)


1.2.2 (doi: 10.5281/zenodo.3921997)
===================================

*Changes*

- obspy.core:
  * Fix wrong values in Stats object after deepcopy or pickle of Stats object
    for edge cases (see #2601)
- obspy.clients.fdsn:
  * EIDA routing client: fix an issue that leaded to a request of *all* EIDA
    data when requesting an invalid, out-of-epochs time window for a valid
    station (see #2611)
  * update RASPISHAKE URL mapping to use https
  * fix a bug of not handling HTTPException in mass_downloader (see #2606)
  * use the client's set timeout in service discovery, too (see #2656)
  * add URL mapping for FDSNWS at UIB-NORSAR (see #2659)
- obspy.clients.filesystem:
  * sds: continue get_all_stations() even if encountering an invalid channel
    code (see #2636)
- obspy.clients.neic:
  * Make client socket blocking (see #2617)
- obspy.io.hypodd:
  * Allow 60.0 seconds in the second field of origin times, allow NaN in
    the magnitude field for pha format (see #2627)
- obspy.io.nordic:
  * Fixed a bug raising an exception when reading a nordic file with a non
    positive-definite covariance matrix (see #2593)
- obspy.io.seiscomp:
  * Fixed an issue where a response has zero zeros or poles (see #2633)


1.2.1
=====
https://doi.org/10.5281/zenodo.3706479


*Changes*

- fix an installation issue with pip and setuptools version 46 (see #2578)
- fix response plots when providing `axes=...` with a numpy array of Axes
  instances (see #2579)


1.2.0
=====
https://doi.org/10.5281/zenodo.3674646

Work on this release was in parts and among others supported by the following
institutions/companies and grants (in alphabetical order):

- Earthquake Commision of New Zealand (EQC), grant 18/753
- École et Observatoire des Sciences de la Terre - Université de Strasbourg
- ETH Zürich
- Friedrich-Schiller-Universität Jena
- Geoscience Australia
- Incorporated Research Institutions for Seismology (IRIS), NSF (SAGE) award
  :: EAR-1851048
- Institut de Physique du Globe de Strasbourg
- Institut de Physique du Globe de Paris
- Institutions for Seismology (IRIS) through the PASSCAL Instrument Center at
  New Mexico Tech. The facilities of the IRIS Consortium are supported by the
  National Science Foundation under Cooperative Agreement EAR-1261681 and the
  DOE National Nuclear Security Administration.
- Istituto Nazionale di Geofisica e Vulcanologia, Osservatorio Etneo (Italy),
  Allegato B2 DPC-INGV 2012-2021 Task 10
- Ludwig-Maximilians-Universität München
- National Institute for Occupational Safety and Health
- Royal Netherlands Meteorological Institute (KNMI)
- School of Geography, Environment and Earth Sciences, Victoria University of
  Wellington
- The European Union’s Horizon 2020 research and innovation programme under
  the ChEESE project, grant agreement No. 823844
- The Royal Observatory of Belgium
- U.S. Geological Survey

*Changes*

- obspy.core
  * Inventory objects have been adapted to StationXML 1.1 for details on
    changes see #2510 and
    https://github.com/FDSN/StationXML/blob/master/Changes.md
  * Fixed import of custom plugins (see #2423)
  * Fixed "==" comparison for Stream and Trace which was very slow in case of
    traces with >1e6 samples (see #2377)
  * Added almost_equal method for Stream and Trace classes (see #2286).
  * Casting FDSN identifiers to strings upon setting in the stats dictionary
    (see #1997).
  * UTCDateTime objects will now always evaluate equal if their string
    representations are equal (see #2049).
  * UTCDateTime objects now issue depreciation warnings when setting any

obspy/signal/src/evalresp/evalresp.c
/* evalresp.c:  Main module for 'evalresp'. */

/*
   10/19/2005 -- [ET]  Added parameters for List-blockette interpolation;
                       added warnings for unrecognized parameters.
    11/3/2005 -- [ET]  Moved 'use_delay()' function from 'calc_fctns.c'
                       to 'evalresp.c'.
     2/6/2006 -- [ET]  Moved 'use_delay()' function from 'evalresp.c'
                       to 'evresp.c'.
    3/28/2006 -- [ET]  Added "free(freqs)" to end of 'main()' function.
*/

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <string.h>
#include <stdlib.h>
#include "evresp.h"

#define DATIMESIZE 32
#define TODAYSIZE 10



int main(int argc, char *argv[])
{
  char *sta_list, *cha_list, units[MAXFLDLEN], *file, *verbose;
  char *net_code, *locid, *type, rtype[MAXFLDLEN];
  int lin_typ = 0, nfreqs, i, fswidx, tmp_val, fldlen;
  int start_stage = -1, stop_stage = 0, stdio_flag = 0;
  int listinterp_out_flag = 0, listinterp_in_flag = 0;
  int unwrap_flag = 0;
  double listinterp_tension = 1000.0;
  char *t_o_day;
  char *datime;
  double incr, freq_lims[2], temp_val, *freqs;
  double val;
  struct response *first;
  char *minfstr, *maxfstr, *numfstr;
  char param_err_msgstr[] = "%s: missing option to argument '%s'";
  int useTotalSensitivityFlag = 0; /* IGD 01/29/10 */
  

  curr_seq_no = -1;

  myLabel[0]= '\0';

  if (argc < 5) {
    printf("EVALRESP V%s\n", REVNUM);
    printf("\nUSAGE: evalresp STALST CHALST YYYY DAY MINFREQ");
    printf(" MAXFREQ NFREQ [options]\n\n");
    printf("  OPTIONS\n\n");
    printf("    '-f file'              (directory-name|filename)\n");
    printf("    '-u units'             ('dis'|'vel'|'acc'|'def')\n");
    printf("    '-t time-of-day'       (HH:MM:SS)\n");
    printf("    '-s type-of-spacing'   (log|lin)\n");
    printf("    '-n netid'             ('II'|'IU'|'G'|'*'...)\n");
    printf("    '-l locid'             ('01'|'AA,AB,AC'|'A?'|'*'...)\n");
    printf("    '-r resp_type'         ('ap'=amp/pha | 'cs'=complex spectra | 'fap'=freq/amp/pha)\n");
    printf("    '-stage start [stop]'  (start and stop are integer stage numbers)\n");
    printf("    '-stdio'               (take input from stdin, output to stdout)\n");
    printf("    '-use-estimated-delay' (use estimated delay instead of correction applied in computation of ASYM FIR response)\n");
    printf("    '-il'                  (interpolate List blockette output)\n");
    printf("    '-ii'                  (interpolate List blockette input)\n");
    printf("    '-it tension'          (tension for List blockette interpolation)\n");
    printf("    '-unwrap'              (unwrap phase if the output is AP) \n");
    printf("    '-ts'                  (Use total sensitivity from stage 0 instead of computed)\n");
    printf("    '-v'                   (verbose; list");
    printf(" parameters on stdout)\n\n");
    printf("    NOTES:\n");
    printf("      (1) If the 'file' argument is a directory, ");
    printf("that directory will be\n          searced for RESP ");
    printf("files of the form RESP.NETID.STA.CHA.  Files\n");
    printf("          of this type are created by rdseed");
    printf(" when it is run with the\n          ");
    printf("'-R' option or when the '-d' option is used and");
    printf(" responses are\n          requested\n"); 
    printf("      (2) If the 'file' argument is a file, that");
    printf(" file is assumed to be\n          output from a ");
    printf("call to rdseed with the '-R' option\n");
    printf("      (3) If the 'file' argument is missing, the");
    printf(" current directory\n          will be searced for ");
    printf("RESP files of the form RESP.NETID.STA.CHA\n");
    printf("      (4) the directory indicated by the environment ");
    printf("variable SEEDRESP\n          will also be searched for");
    printf(" the requested files");
    printf(" (if it is defined),\n          but if matching station-");
    printf("channel-network values are found in\n");
    printf("          both directories, then the local files");
    printf(" take precedence\n");
    printf("      (5) the NETID (above) indicates a one or two letter network code.\n");
    printf("          If no network code exists in the SEED volume, a value of ''\n");
    printf("          or '*' will match the 'no network' case (since the ?? string\n");
    printf("          in the RESP file output is replaced with an empty string in\n");
    printf("          the program).\n");
    printf("      (6) the '-stage' option allows the user to specify the range of\n");
    printf("          stages that will be included in the calculation.  If only one\n");
    printf("          stage number is given, a response will only be calculated for\n");
    printf("          that stage.  If both a start and stop stage number are given,\n");
    printf("          any stage between (and including) the start and stop stages\n");
    printf("          will be included in the calculation.\n\n");
    printf("  EXAMPLES:\n");
    printf("evalresp AAK,ARU,TLY VHZ 1992 21 0.001 10 100 -f ");
    printf("/sd15/EVRESP/NEW/rdseed.out\n");
    printf("evalresp KONO BHN,BHE 1992 1 0.001 10 100 -f ");
    printf("/sd15/EVRESP/NEW -t 12:31:04 -v\n");
    printf("evalresp FRB BHE,BHZ 1994 31 0.001 10 100 -f ");
    printf("resp.all_stations -n '*' -v\n\n");    exit(1);
  }

    /* find index of first switch parameter */
  fswidx = 0;                /* loop until switch param found (if any) */
  while(++fswidx < argc &&   /*  (check if switch param or real #) */
               (strncmp(argv[fswidx],"-",1) != 0 || is_real(argv[fswidx])));
  if(fswidx < 5) {
    error_exit(USAGE_ERROR,"Not all of the required inputs are \n\t"
            "present (%d missing), type '%s' for usage", 8-fswidx, argv[0]);
  }
    /* setup min,max,num freq values; use defaults if missing */
  minfstr = (fswidx > 5) ? argv[5] : "1.0";
  maxfstr = (fswidx > 6) ? argv[6] : minfstr;
  numfstr = (fswidx > 7) ? argv[7] : "1";

    /* warn about any unexpected non-switch parameters */
  for(i=8; i<fswidx; ++i)
  {
    if(argv[i] != NULL && argv[i][0] != '\0')
      fprintf(stderr,"WARNING:  Unrecognized parameter:  %s\n",argv[i]);
  }

  /* initialize the optional arguments */

  /* If user did not define -use-delay option by default it is FALSE */
  use_delay(FALSE);

  strncpy(units,"",MAXFLDLEN);
  strncpy(rtype,"",MAXFLDLEN);
  net_code = locid = file = verbose = type = (char *)NULL;
  if((t_o_day = (char *)malloc(strlen("00:00:00")+1)) == (char *)NULL) {
  }
  sprintf(t_o_day, "00:00:00");

  /* then get the optional arguments (if any) */

  for (i = fswidx; i < argc; i++) {
    if (0 == strcmp(argv[i], "-use-estimated-delay"))
       use_delay(TRUE);
    else if (0 == strcmp(argv[i], "-use-delay"))  /* IGD 04/29/09 Backward compatibility */
       use_delay(TRUE);
    else if(!strcmp(argv[i], "-u")){
      if((++i) < argc && *argv[i] != '-')
        strncpy(units,argv[i],MAXFLDLEN);
      else
        error_exit(USAGE_ERROR,param_err_msgstr,argv[0],argv[i-1]);
    }
    else if(!strcmp(argv[i], "-r")){
      if((++i) < argc && *argv[i] != '-')
        strncpy(rtype,argv[i],MAXFLDLEN);
      else
        error_exit(USAGE_ERROR,param_err_msgstr,argv[0],argv[i-1]);
    }
    else if(!strcmp(argv[i], "-stdio")){
      stdio_flag = 1;
    }
    else if(!strcmp(argv[i], "-stage")){
      if((++i) < argc && is_int(argv[i]))
        start_stage = atoi(argv[i]);
      else
        error_exit(USAGE_ERROR,param_err_msgstr,argv[0],argv[i-1]);
      if((++i) < argc && is_int(argv[i])) {
        tmp_val = atoi(argv[i]);
        if(tmp_val > start_stage)
          stop_stage = tmp_val;
        else {
          stop_stage = start_stage;
          start_stage = tmp_val;
        }
      }
      else {
        i--;
      }
    }
    else if(!strcmp(argv[i], "-n")){
      if((++i) < argc && *argv[i] != '-')
        net_code = argv[i];
      else
        error_exit(USAGE_ERROR,param_err_msgstr,argv[0],argv[i-1]);
    }
    else if(!strcmp(argv[i], "-l")){
      if((++i) < argc && *argv[i] != '-')
        locid = argv[i];
      else
        error_exit(USAGE_ERROR,param_err_msgstr,argv[0],argv[i-1]);
    }
    else if(!strcmp(argv[i], "-t")){
      if((++i) < argc && *argv[i] != '-') {
        fldlen = strlen(argv[i]);
        if(fldlen >= TODAYSIZE) {
            t_o_day = realloc(t_o_day,(fldlen+1));
        }
        strcpy(t_o_day, argv[i]);
      }
      else
        error_exit(USAGE_ERROR,param_err_msgstr,argv[0],argv[i-1]);
    }
    else if(!strcmp(argv[i], "-f")){
      if((++i) < argc && *argv[i] != '-')
        file = argv[i];
      else
        error_exit(USAGE_ERROR,param_err_msgstr,argv[0],argv[i-1]);
    }
    else if(!strcmp(argv[i], "-s")){
      if((++i) < argc && *argv[i] != '-')
        type = argv[i];
      else
        error_exit(USAGE_ERROR,param_err_msgstr,argv[0],argv[i-1]);
      if(strcmp(type,"lin") & strcmp(type,"log")){
        error_exit(USAGE_ERROR,"%s: option '-s' illegal type '%s'",
                   argv[0], type);
      }
    }
    else if(!strcmp(argv[i], "-unwrap"))
      unwrap_flag = 1;
    else if(!strcmp(argv[i], "-ts"))
      useTotalSensitivityFlag = 1;
    else if(!strcmp(argv[i], "-il"))
      listinterp_out_flag = 1;
    else if(!strcmp(argv[i], "-ii"))
      listinterp_in_flag = 1;
    else if(!strcmp(argv[i], "-it")){
      if((++i) < argc && *argv[i] != '-') {
        if(is_real(argv[i]))
          listinterp_tension = atof(argv[i]);
        else {
          error_exit(USAGE_ERROR,
              "%s:  illegal value for 'it' parameter:  %s",argv[0],argv[i]);
        }
      }
      else
        error_exit(USAGE_ERROR,param_err_msgstr,argv[0],argv[i-1]);
    }
    else if(!strcmp(argv[i], "-v"))
      verbose = argv[i];
    else if(argv[i] != NULL && argv[i][0] != '\0')
      fprintf(stderr,"WARNING:  Unrecognized parameter:  %s\n",argv[i]);
  }

  sta_list = argv[1];
  cha_list = argv[2];

  if(!is_int(argv[3]))
    error_exit(USAGE_ERROR,"year must be an integer, found '%s'", argv[3]);
  if(!is_int(argv[4]))
    error_exit(USAGE_ERROR,"julian day must be an integer, found '%s'", argv[4]);
  if(!is_time(t_o_day))
    error_exit(USAGE_ERROR,"'time of day' must have format 'hh[:mm[:ss[.#####]]]', found '%s'",
                           t_o_day);

  fldlen = strlen(argv[3])+strlen(argv[4])+strlen(t_o_day)+3;
  if((datime = (char *)malloc(fldlen)) == (char *)NULL) {
  }
  sprintf(datime, "%s,%s,%s", argv[3], argv[4], t_o_day);

  if(!is_real(minfstr) || !is_real(maxfstr))
    error_exit(USAGE_ERROR,"freq_lims must be real numbers, found (%s,%s)",
               minfstr,maxfstr);
  freq_lims[0] = atof(minfstr);
  freq_lims[1] = atof(maxfstr);
  if ((freq_lims[0] == 0.0 || freq_lims[1] == 0.0) &&
                                (type == NULL || strcmp(type,"lin") != 0)) {
    error_exit(USAGE_ERROR," freq lims can't equal 0 if log spacing is used");
  }
  if(!is_int(numfstr))
    error_exit(USAGE_ERROR,"nfreqs must be an integer, found '%s'", numfstr);
  nfreqs = atoi(numfstr);

  /* check the frequency input values and range */

  if(freq_lims[1] < freq_lims[0]){
    temp_val = freq_lims[1];
    freq_lims[1] = freq_lims[0];
    freq_lims[0] = temp_val;
  }
  else if(nfreqs <= 0){
    error_exit(USAGE_ERROR,"nfreqs value entered must be positive, entered %d",
               nfreqs);
  }

  /* depending on the type of spacing, set either a step size or a multiplication
     factor for the loop over frequencies */

  if(type != NULL && !strcmp(type,"lin")){
    lin_typ = 1;
    if (nfreqs == 1)
      incr = -1.0;
    else
      incr = (freq_lims[1]-freq_lims[0])/(nfreqs-1);
  }
  else
    if (nfreqs == 1)
      incr = -1.0;
    else
      incr = pow((freq_lims[1]/freq_lims[0]), (1./(nfreqs-1)));

/* if net_code not given, default to "*" value */

  if(net_code == (char *)NULL)
    net_code = "*";

/* if locid not given, default to "*" value */

  if(locid == (char *)NULL)
    locid = "*";

/* convert the units and response type arguments to lower case after
   checking to make sure they match one of the appropriate input values */

  if(strlen(rtype)) {
    for(i = 0; i < (int)strlen(rtype); i++)
      *(rtype+i) = toupper(*(rtype+i));
    if(strcmp(rtype,"CS") == 0)        /* if complex-spectra output then */
      listinterp_out_flag = 0;         /* force List-out-interp flag clear */
    else if((strcmp(rtype,"AP") != 0)) {  /* if invalid value then abort */
      if ((strcmp(rtype,"FAP") != 0))
	error_exit(USAGE_ERROR,"evalresp; rtype entered ('%s') not a recognized string (see usage)", rtype);
    }
  }
  else
    strncpy(rtype,"AP",MAXFLDLEN);

  if(strlen(units)) {
    for(i = 0; i < (int)strlen(units); i++)
      *(units+i) = toupper(*(units+i));
    if(strcmp(units,"DIS") && strcmp(units,"VEL") && strcmp(units,"ACC") && strcmp(units,"DEF"))
      error_exit(USAGE_ERROR,"evalresp; units entered ('%s') not a recognized string (see usage)",
                 units);
    else if(!strcmp(units,"DEF"))
      def_units_flag = 1;
    else
      def_units_flag = 0;
  }
  else
    strncpy(units,"VEL",MAXFLDLEN);

/* allocate space for the frequencies and fill with appropriate values */

  freqs = alloc_double(nfreqs);
  for(i = 0, val = freq_lims[0]; i < nfreqs; i++) {
    freqs[i] = val;
    if(lin_typ)
      val += incr;
    else
      val *= incr;
  }


/* then get the response for each of the requested station-channel pairs at
    each of the requested frequencies */

  first = evresp_itp(sta_list,cha_list,net_code,locid,datime,units,file,
                     freqs,nfreqs,rtype,verbose,start_stage,stop_stage,
                     stdio_flag,listinterp_out_flag,listinterp_in_flag,
                     listinterp_tension, useTotalSensitivityFlag);
  if (!first)
  {
	fprintf(stderr, "EVRESP FAILED\n");
	exit(-1);
  }

/* and print the responses to a set of files */

  print_resp_itp(freqs,nfreqs,first,rtype,stdio_flag,
                                    listinterp_out_flag,listinterp_tension, unwrap_flag);

  free_response(first);
  free(freqs);                    /* added 3/28/2006 -- [ET] */

  exit(0);
  return 0;             /* 'return' statement to avoid compiler warning */
}

